pkg/prom/cluster/node.go

```diff
@@ -0,0 +1,327 @@
+
+		firstError error
+	)
+
+	backoff := cortex_util.NewBackoff(ctx, backoffConfig)
+	for backoff.Ongoing() {
+		rs, err = n.ring.GetAllHealthy(ring.Read)
+		if err == nil {
+			break
+		}
+		backoff.Wait()
+	}
+	if err := backoff.Err(); err != nil && firstError == nil {
+		firstError = err
+	}
+
+	if len(rs.Ingesters) > 0 {
+		level.Info(n.log).Log("msg", "informing remote nodes to reshard")
+	}
+
+	_, err = rs.Do(ctx, 500*time.Millisecond, func(c context.Context, id *ring.InstanceDesc) (interface{}, error) {
+		// Skip over ourselves.
+		if id.Addr == n.lc.Addr {
+			return nil, nil
+		}
+
+		ctx = user.InjectOrgID(ctx, "fake")
+		return nil, n.notifyReshard(ctx, id)
+	})
+	if err != nil && firstError == nil {
+		firstError = err
+	}
+
+	if includeSelf {
+		level.Info(n.log).Log("msg", "running local reshard")
+		if _, err := n.srv.Reshard(ctx, &pb.ReshardRequest{}); err != nil {
+			level.Warn(n.log).Log("msg", "dynamic local reshard did not succeed", "err", err)
+		}
+	}
+
+	return firstError
+}
+
+// notifyReshard informs an individual node to reshard.
+func (n *node) notifyReshard(ctx context.Context, id *ring.InstanceDesc) error {
+	cli, err := client.New(n.cfg.Client, id.Addr)
+	if err != nil {
+		return err
+	}
+	defer cli.Close()
+
+	level.Info(n.log).Log("msg", "attempting to notify remote agent to reshard", "addr", id.Addr)
+
+	backoff := cortex_util.NewBackoff(ctx, backoffConfig)
+	for backoff.Ongoing() {
+		_, err := cli.Reshard(ctx, &pb.ReshardRequest{})
+		if err == nil {
+			break
+		}
+
+		level.Warn(n.log).Log("msg", "reshard notification attempt failed", "addr", id.Addr, "err", err, "attempt", backoff.NumRetries())
+		backoff.Wait()
+	}
+
+	return backoff.Err()
+}
+
+// Stop stops the node and cancels it from running. The node cannot be used
+// again once Stop is called.
+func (n *node) Stop() error {
+	n.mut.Lock()
+	defer n.mut.Unlock()
+
+	if n.exited {
+		return fmt.Errorf("node already exited")
+	}
+	n.exited = true
+
+	level.Info(n.log).Log("msg", "shutting down node")
+
+	// Shut down dependencies. The lifecycler *MUST* be shut down first since n.ring is
+	// used during the shutdown process to inform other nodes to reshard.
+	//
+	// Note that stopping the lifecycler will call performClusterReshard and will block
+	// until it completes.
+	var firstError error
+	for _, dep := range []services.Service{n.lc, n.ring} {
+		if dep == nil {
+			continue
+		}
+		err := services.StopAndAwaitTerminated(context.Background(), dep)
+		if err != nil && firstError == nil {
+			firstError = err
+		}
+	}
+
+	close(n.reload)
+	level.Info(n.log).Log("msg", "node shut down")
+	return firstError
+}
+
+// Flush implements ring.FlushTransferer. It's a no-op.
+func (n *node) Flush() {}
+
+// TransferOut implements ring.FlushTransferer. It connects to all other healthy agents and
+// tells them to reshard. TransferOut should NOT be called manually unless the mutex is
+// held.
+func (n *node) TransferOut(ctx context.Context) error {
```

Comment:
Is there a reason we don't hold the mutex here?

Reply:
Yep, if we hold the mutex here then it deadlocks with `ApplyConfig` when stopping the previous lifecycler 🙃. I'm not thrilled about this, but I can't think of a better way of handling it.

Reply:
Brainstorming, since this looks like it is only called when leaving the ring and not called again under normal use can we use [Once](https://golang.org/pkg/sync/#Once) to help in anyway?

Reply:
Nah, that would normally help but not in this case. Every time you call `ApplyConfig` you will leave the old ring from the old config before joining the new ring with the new config, so `sync.Once` would break here.